Add ruler file for customizing the ruler

joelim-work opened the pull request:
## Related

- Supersedes #2083
- Fixes #2057
- Closes #2084
- Closes #2120

## Description

This PR introduces the ruler file as an alternative to `rulerfmt`/`statfmt` for customizing the ruler at the bottom of the UI. The syntax of the ruler file is a [Go template](https://pkg.go.dev/text/template), which allows for more powerful customization, including access to conditional logic and `printf` formatting features.

Although this feature is intended to eventually replace the existing `rulerfmt`/`statfmt` options, it is also a large change (`text/template` is added as a dependency, and a new `ruler` config file is introduced), so a more gradual rollout strategy will be used:

1. Support for the `ruler` file will be added alongside `rulerfmt`/`statfmt`, and will be disabled by default. Use `set rulerfile true` to enable the `ruler` file. At this point the `ruler` file shall be considered as an experimental feature.
2. The `ruler` file feature will have its experimental status lifted, and then be enabled by default. At this point in time users can still enable the `rulerfmt`/`statfmt` by using `set rulerfile false`.
3. The `rulerfmt`/`statfmt` options will be marked as deprecated.
4. The `rulerfmt`/`statfmt` options will be removed.

As an experimental feature, if there is significant pushback from users or other issues, the `ruler` file will be rejected with all changes reverted.

## Examples

Default ruler:

```handlebars
{{with .Message -}}
    {{. -}}
{{else with .Stat -}}
    {{.Permissions | printf "\033[36m%s\033[0m" -}}
    {{with .LinkCount}} {{.}}{{end -}}
    {{with .User}} {{.}}{{end -}}
    {{with .Group}} {{.}}{{end -}}
    {{.Size | humanize | printf " %5s" -}}
    {{.ModTime | printf " %s" -}}
    {{with .Target}} -> {{.}}{{end -}}
{{end -}}
{{.SPACER -}}
{{with .Keys}}  {{.}}{{end -}}
{{with .Progress}}  {{join . " "}}{{end -}}
{{with .Copy}}  {{len . | printf "\033[7;33m %d \033[0m"}}{{end -}}
{{with .Cut}}  {{len . | printf "\033[7;31m %d \033[0m"}}{{end -}}
{{with .Select}}  {{len . | printf "\033[7;35m %d \033[0m"}}{{end -}}
{{with .Visual}}  {{len . | printf "\033[7;36m %d \033[0m"}}{{end -}}
{{with .Filter}}  {{join . " " | printf "\033[7;34m %s \033[0m"}}{{end -}}
{{printf "  %d/%d" .Index .Total}}
```

Display options:

```handlebars
hidden={{.Options.hidden}}
```

Display user options:

> [!NOTE]  
> Add `set user_foo bar` to the `lfrc` file.

```handlebars
foo={{.UserOptions.foo}}
```

Display environment variables:

```handlebars
LF_LEVEL={{env "LF_LEVEL"}}
```

A single `{{.SPACER}}` can be used to split into left and right sections:

```handlebars
left{{.SPACER}}right
```

Padding with spaces:

```handlebars
{{printf "[%-8s][%8s]" "left" "right"}}
```

Colors and styles can be applied using terminal escape sequences:

```handlebars
{{printf "\033[4;31m%s\033[0m" "hello world"}}
```

Conditional logic:

```handlebars
{{with .Stat -}}
    {{.Path}}{{if .Target}} -> {{.Target}}{{end}}
{{else -}}
    empty directory
{{end -}}
```

Different colors for modes:

```handlebars
{{$color := 32 -}}
{{if eq .Mode "VISUAL" -}}
    {{$color = 34 -}}
{{end -}}

{{printf "\033[7;%dm %s \033[0m" $color .Mode}}
```

Different colors for file permissions:

```handlebars
{{with .Stat -}}
    {{printf "\033[34m%s" (substr .Permissions 0 1) -}}
    {{printf "\033[33m%s" (substr .Permissions 1 1) -}}
    {{printf "\033[31m%s" (substr .Permissions 2 1) -}}
    {{printf "\033[32m%s" (substr .Permissions 3 1) -}}
    {{printf "\033[33m%s" (substr .Permissions 4 1) -}}
    {{printf "\033[31m%s" (substr .Permissions 5 1) -}}
    {{printf "\033[32m%s" (substr .Permissions 6 1) -}}
    {{printf "\033[33m%s" (substr .Permissions 7 1) -}}
    {{printf "\033[31m%s" (substr .Permissions 8 1) -}}
    {{printf "\033[32m%s" (substr .Permissions 9 1) -}}
    {{printf "\033[0m"}}
{{end -}}
```

Comments are allowed:

```handlebars
{{/* this is a comment */ -}}
this will be shown
```

joelim-work commented:
> This is huge. I am interested in the user responses (I hope we even get some).

So just to be clear, this PR isn't any more 'huge' than the previous one #2083, it's pretty much the same code but I decided to existing `rulerfmt`/`statfmt` options for now instead of replacing them.

CatsDeservePets commented:
Yeah, my comment was meant to stand on its own as if the other PR never existed.

I am definitely exited and give it a deeper look within the next days.

joelim-work commented:
Oh, I was under the impression you already looked at this in detail a long time ago. I don't mind leaving this PR open for a while.

CatsDeservePets reviewed (commented):
[See my other comment]

joelim-work commented on ui.go:
```diff
@@ -1016,6 +1020,126 @@ func (ui *ui) drawRuler(nav *nav) {
+		return
+	}
+
+	ui.msgWin.print(ui.screen, 0, 0, tcell.StyleDefault, left)
```
This is fixed now, I decided to just expand all occurrences of `{{.SPACER}}` with the appropriate number of spaces. The following should work now:

```handlebars
{{printf "\033[41m"}}hello{{.SPACER}}world{{printf "\033[0m"}}
```

joelim-work commented:
Thanks for your comments.

I have changed the behavior of error-reporting so that it is shown directly in the UI (a badly-formed template will show an error instead of continuing to the next valid file), and also added support for multiple spacers. Both of these changes made the code simpler in some way so I think they are good to have.

Regarding the use of a separate config file, the reason for this is that ruler configuration has become sufficiently complex that using a string option is no longer suitable. You are right that it is loaded only once on startup, however the same problem exists for the colors and icons files too - so far I haven't seen any user requests about wanting to change them dynamically while `lf` is running. If you just wanted to specify an alternate ruler file (e.g. when running in a different environment, you could try setting `LF_CONFIG_PATH` to effectively have a different config profile. It might also be possible to just place all the logic for different profiles inside the one ruler file itself.

As for supporting other functions like string manipulation and arithmetic operations, there is no good solution since Go templates is just a templating language but you are expecting it to have the functionality of a proper scripting language. I do not mind exporting a few functions here and there (based on user-demand), but there is practically no limit as to what can be requested. I have decided to just export `%p` (line percentage) directly since Vim does this as well.

This does leave the question of "why not just use Lua?" if Go templates are still too limited. But if we use Lua to configure the ruler, then we might as well use Lua to configure everything, and at that point it is probably easier to just write a brand new file manager from scratch (a.k.a `yazi`).

CatsDeservePets commented:
> I have changed the behavior of error-reporting so that it is shown directly in the UI (a badly-formed template will show an error instead of continuing to the next valid file), and also added support for multiple spacers. Both of these changes made the code simpler in some way so I think they are good to have.

That's great. Thanks to these changes, changing the background colour/style of the `ruler` now works flawlessly. I really like that.

> Regarding the use of a separate config file, the reason for this is that ruler configuration has become sufficiently complex that using a string option is no longer suitable. You are right that it is loaded only once on startup, however the same problem exists for the colors and icons files too - so far I haven't seen any user requests about wanting to change them dynamically while lf is running. If you just wanted to specify an alternate ruler file (e.g. when running in a different environment, you could try setting LF_CONFIG_PATH to effectively have a different config profile. It might also be possible to just place all the logic for different profiles inside the one ruler file itself.

I am fully aware why this has been done. Keeping this config a single string var would be a hell to escape properly (and not to mention the line width of the config file). I just wanted to pinpoint potential breaking changes with this.

> As for supporting other functions like string manipulation and arithmetic operations, there is no good solution since Go templates is just a templating language but you are expecting it to have the functionality of a proper scripting language. I do not mind exporting a few functions here and there (based on user-demand), but there is practically no limit as to what can be requested. I have decided to just export %p (line percentage) directly since Vim does this as well.
> 
> This does leave the question of "why not just use Lua?" if Go templates are still too limited. But if we use Lua to configure the ruler, then we might as well use Lua to configure everything, and at that point it is probably easier to just write a brand new file manager from scratch (a.k.a yazi).

Those are my exact thoughts. It's kind of a dilemma. How much of this will actually end up being used by others? At which point does `lf` stop begin lightweight and start begin bloated? I don't know. I want to keep @gokcehan's vision alive (if it's not too late already).

I also stumbled across https://github.com/Masterminds/sprig which seems capable of more in-depth string manipulation and arithmetic operations. However, I am kinda opposed to adding more (third party) dependencies.

Nevertheless, I think this template-based ruler offers lots of benefits in terms of customisation. 

-----

One more thought: 
The way `lf` handles rulers that exceed the available space could be improved.

| enough space | not enough space |
|--------------|------------------|
| <img width="1033" height="801" src="https://github.com/user-attachments/assets/1893fef6-62a6-4e0b-97e9-3ffed3c370ec" /> | <img width="1033" height="801" src="https://github.com/user-attachments/assets/c75d88a6-d57b-4fbc-9893-53808d14fa10" /> |

At the bare minimum, `truncatechar` should be inserted in my opinion. `Vim`'s statusline also allows defining the position by placing a special expansion (even multiple times for different parts of the statusline):
```
	< -   Where to truncate line if too long.  Default is at the start.
	      No width fields allowed.
```
`lf` on the other hand, already supports `truncatepct`. Could it be useful to use it here as well?

joelim-work commented on etc/ruler.default:
```diff
@@ -0,0 +1,20 @@
+{{.SPACER -}}
+{{with .Keys}}  {{.}}{{end -}}
+{{with .Progress}}  {{join . " "}}{{end -}}
+{{with .Copy}}  {{len . | printf "\033[7;33m %d \033[0m"}}{{end -}}
```
I guess this makes things more convenient when configuring `copyfmt`/`cutfmt`/`selectfmt`/visualfmt`. I have made the change as you suggested.

gokcehan commented:
@CatsDeservePets I saw your email today. It's great to hear that @joelim-work sees you as good fit as a new collaborator. I have invited you as a collaborator for the repository now. Thank you very much for your interest in taking initiative and welcome to the team!

As you know, I haven't been around for quite a while due to lack of time and interest. I mostly stopped following the development except for checking out the commit history and release notes every once in a while. However, I should say that you have been doing a great job as maintainers and I'm still a happy user of `lf`. Please feel free to let me know if you need anything from me again.

joelim-work commented:
> One more thought. How about, instead of `rulerfile` being a boolean, let it be a path string instead (like `previewer` and `cleaner`). This would allow using a different ruler file "on the fly", solving the problem of quickly swapping out the ruler.

That is pretty much a feature request (i.e. changing ruler file dynamically). Currently the ruler file is like the colors/icons files which are at a fixed location and loaded once on startup. Allowing it to be changed dynamically requires adding a new option to specify the path.

Also note that `previewer` and `cleaner` are not necessarily paths, they are the name of executable programs which can use `PATH` as part of the lookup. That is why they are not stored in the same way as the colors/icons files.
